### PR TITLE
ACL Enhancements

### PR DESCRIPTION
**Pull request by softcraftsman**
Not ready for a PR merge, just building a new website.

### CODE REVIEW

**Comment by SvenAelterman** [comment]
@softcraftsman Is this ready for review yet?

File: src/api/Endpoints/FileSystems.cs [hunk 1]
```diff
@@ -168,7 +152,7 @@ private static async Task<IActionResult> FileSystemsPOST(HttpRequest req, ILogge
                 return new BadRequestErrorMessageResult("Owner identity not found. Please verify that the Owner is a valid member UPN and that the application has User.Read.All permission in the directory.");
 
             // Call each of the steps in order and error out if anytyhing fails
-            var storageUri = new Uri($"https://{tlfp.StorageAcount}.dfs.core.windows.net");
+            var storageUri = SasConfiguration.GetStorageUri("tlfp.StorageAcount");
```
Comment (SvenAelterman) [inline src/api/Endpoints/FileSystems.cs:155]: I don't understand how this works with quotes around `tlfp.StorageAccount`?

File: src/api/Endpoints/TopLevelFolders.cs [hunk 2]
```diff
@@ -98,27 +103,36 @@ public static class TopLevelFolders
...
-			var storageUri = new Uri($"https://{tlfp.StorageAcount}.dfs.core.windows.net");
+			Result result = null;
+			var storageUri = SasConfiguration.GetStorageUri(account);
 			var fileSystemOperations = new FileSystemOperations(log, new DefaultAzureCredential(), storageUri);
 			var folderOperations = new FolderOperations(log, new DefaultAzureCredential(), storageUri, tlfp.FileSystem);
 
-			Result result = null;
-			result = await folderOperations.CreateNewFolder(tlfp.Folder);
+			// Add folder owner or a larger group to the container ACL
+			var groupACL = Environment.GetEnvironmentVariable("ACL_ROOT_GROUP");
```
Comment (SvenAelterman) [inline src/api/Endpoints/TopLevelFolders.cs:112]: I get the concept here, but I doubt they're will be one system-wide root group. Perhaps one per storage account?

File: src/api/Endpoints/TopLevelFolders.cs [hunk 3]
```diff
@@ -98,27 +103,36 @@ public static class TopLevelFolders
...
+			// Folder Metadata
 			result = await folderOperations.AddMetaData(tlfp.Folder, tlfp.FundCode, tlfp.FolderOwner);
 			if (!result.Success)
 				return new BadRequestErrorMessageResult(result.Message);
 
-			result = await folderOperations.AssignFullRwx(tlfp.Folder, tlfp.FolderOwner);
+			// Foler permissions
+			if (!tlfp.UserAccessList.Contains(tlfp.FolderOwner))
+				tlfp.UserAccessList.Add(tlfp.FolderOwner);
+			result = await folderOperations.AssignFullRwx(tlfp.Folder, tlfp.UserAccessList);
```
Comment (SvenAelterman) [inline src/api/Endpoints/TopLevelFolders.cs:131]: Add whitespace for clarity

File: src/api/Endpoints/TopLevelFolders.cs [hunk 4]
```diff
@@ -135,8 +149,12 @@ internal static async Task<TopLevelFolderParameters> GetTopLevelFolderParameters
 					log.LogError("Body was empty coming from ReadToEndAsync");
 				}
 			}
-			var bodyDeserialized = JsonConvert.DeserializeObject<TopLevelFolderParameters>(body);
-			return bodyDeserialized;
+			var tlfp = JsonConvert.DeserializeObject<TopLevelFolderParameters>(body);
+
+			if (tlfp.UserAccessList == null)
+				tlfp.UserAccessList = new List<string>();
```
Comment (SvenAelterman) [inline src/api/Endpoints/TopLevelFolders.cs:155]: Achieve this by putting an initializer in the class definition?

File: src/api/Services/FolderOperations.cs [hunk 5]
```diff
@@ -16,11 +18,10 @@ internal class FolderOperations
 		private readonly ILogger log;
 		private readonly DataLakeFileSystemClient dlfsClient;
 		private readonly decimal costPerTB;
-		private readonly TokenCredential tokenCredential;
+
 		public FolderOperations(ILogger log, TokenCredential tokenCredential, Uri storageUri, string fileSystem)
```
Comment (SvenAelterman) [inline src/api/Services/FolderOperations.cs:22]: Why is `TokenCredential` still being injected in the ctor?

**Review by SvenAelterman — CHANGES_REQUESTED** [review]

File: src/api/Services/FolderOperations.cs (hunk 5 above)
Comment (softcraftsman) [reply src/api/Services/FolderOperations.cs:22]: By injecting it in the constructor, every call within the instance object will be of the correct context. If you wish to create a second instance in another context, you can.

Comment (SvenAelterman) [reply src/api/Services/FolderOperations.cs:22]: Sure, that makes sense, but the instance-level object isn't there anymore. So the ctor gets it, but it just goes to a black hole?

**Comment by SvenAelterman** [comment]
@softcraftsman See refactorings and code comments in acl-enhancements-review branch. Suggest squash merging this branch in acl-enhancements first.

**Review by softcraftsman — COMMENTED** [review]
I believe we have reviewed and approved this massive change enough. Please merge.

**Pull request merged**